### ISSUE TITLE
Should obtain SA token from token secret

### ISSUE THREAD

**Issue by erikgb** [maxgio92/proxy-kubeconfig-generator#3]
This tool does not seem to work on Openshift, as it attempts to obtain the SA token from the Openshift registry image pull secret. This is an automated secret created for all SA in Openshift clusters. I think the algorithm to find the correct secret (containing a token) must be improved.

````
$ proxy-kubeconfig-generator --kubeconfig-secret-key kubeconfig --namespace erikbo --server 'https://capsule-proxy.capsule-system.svc:9001' --server-tls-secret-namespace capsule-system --server-tls-secret-name capsule-proxy --serviceaccount gitops-reconciler
panic: secret gitops-reconciler-dockercfg-qv5nr does not contain a token

goroutine 1 [running]:
main.main()
        /home/erikbo/go/pkg/mod/github.com/maxgio92/proxy-kubeconfig-generator@v0.1.0/main.go:68 +0x434
````

Ref. https://capsule.clastix.io/docs/guides/flux2-capsule/#the-recipe

**Comment by maxgio92** [comment]
Hi @erikgb, thank you for reporting this.

TL;DR this tool doesn't support as of now Kubernetes >= 1.24.

From Kubernetes 1.24 Service Account token are no longer automatically generated, when a Service Account is created ([docs](https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/#manual-secret-management-for-serviceaccounts))
This tool instead assumes a Service Account token exists as a `Secret` object, related to the `ServiceAccount` object. 

I'm sorry this is not stated clearly in the readme.

Anyway, we can think of a better way to manage this situation, if possible, without introducing breaking changes.

**Comment by maxgio92** [comment]
In any case, this is a non-well managed error, as it would be better to not panic, thank you.

**Comment by erikgb** [comment]
> From Kubernetes 1.24 Service Account token are no longer automatically generated, when a Service Account is created ([docs](https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/#manual-secret-management-for-serviceaccounts)) This tool instead assumes a Service Account token exists as a `Secret` object, related to the `ServiceAccount` object.

Even if Kubernetes stopped generating SA token secrets, OpenShift still do. We are on OpenShift 4.12 (Kubernetes 1.25). The secret with the token is there, but there is another secret that seems to confuse the tool. I think it should be fixed by looking for the correct secret containing the token. And if no such secret exists, it should error out - of course.

**Comment by erikgb** [comment]
If it looks for secrets of the correct type (`kubernetes.io/service-account-token`), this should be fixed:

````yaml
kind: Secret
apiVersion: v1
metadata:
  name: gitops-reconciler-token-hsglk
  namespace: erikbo
  annotations:
    kubernetes.io/created-by: openshift.io/create-dockercfg-secrets
    kubernetes.io/service-account.name: gitops-reconciler
    kubernetes.io/service-account.uid: 8d649e38-54dd-413c-811a-8530b950f4a1
data:
  token: ...
type: kubernetes.io/service-account-token
````

**Comment by maxgio92** [comment]
I wasn't aware of OpenShift logics, but if this tool can support OpenShift, that's good :-)

You're right. I just opened an issue to track this other bug.

**Comment by maxgio92** [comment]
Hey @erikgb, in the meantime, I just released a [v0.1.1](https://github.com/maxgio92/proxy-kubeconfig-generator/releases/tag/v0.1.1) that addresses the minor issues you highlighted above.

You can also use it [in-container](https://github.com/maxgio92/proxy-kubeconfig-generator/pkgs/container/proxy-kubeconfig-generator).
And for both check either SBOMs and signatures.

**Comment by erikgb** [comment]
Hi, @maxgio92! It still doesn't do the job. You seem to have fixed the panic, but still not able to do the job correctly - since it is trying to obtain the token from the wrong secret.

````
$ proxy-kubeconfig-generator --kubeconfig-secret-key kubeconfig --namespace erikbo --server 'https://capsule-proxy.capsule-system.svc:9001' --server-tls-secret-namespace capsule-system --server-tls-secret-name capsule-proxy --serviceaccount gitops-reconciler

error generating kubeconfig from service account: secret gitops-reconciler-dockercfg-q46sx does not contain a token
````

````
$ kubectl get secret -n erikbo | grep gitops-reconciler
gitops-reconciler-dockercfg-q46sx   kubernetes.io/dockercfg               1      24h
gitops-reconciler-token-hsglk       kubernetes.io/service-account-token   4      24h
````

**Comment by maxgio92** [comment]
Thanks for reporting it @erikgb

**Comment by maxgio92** [comment]
Just released [v0.1.2](https://github.com/maxgio92/proxy-kubeconfig-generator/releases/tag/v0.1.2) @erikgb :-)

**Comment by erikgb** [comment]
Still no luck, I will take a look at the code. ;-)

````
$ proxy-kubeconfig-generator --kubeconfig-secret-key kubeconfig --namespace erikbo --server 'https://capsule-proxy.capsule-system.svc:9001' --server-tls-secret-namespace capsule-system --server-tls-secret-name capsule-proxy --serviceaccount gitops-reconciler
error generating kubeconfig from service account: secret of type service-account-token not found
````

**Comment by maxgio92** [comment]
@erikgb is the `secrets` field of the `ServiceAccount` _gitops-reconciler_ filled by the bindings to the `Secret`s?

Would you mind running:
```
kubectl get serviceaccount -n erikbo gitops-reconciler -o=yaml
```

**Comment by erikgb** [comment]
I am pretty sure that is the problem. This field is not supposed to be used for this, at least as I read the API docs: https://github.com/kubernetes/api/blob/dc63998f1312927aa27e36e646ddf504470afd14/core/v1/types.go#L4965-L4973

Even if I create a dedicated SA token secret using https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/#create-token, it is not listed on the service account. So I think we should list and filter the secrets in the namespace.
